Python Datetime
--

Python Dates
--
A date in Python is not a data type of its own, but we can import a module named datetime to work with dates as date objects.

In [1]:
# Import the datetime module and display the current date:

import datetime

x = datetime.datetime.now()
print(x)

2023-02-22 21:34:37.472916


Date Output
--
When we execute the code from the example above the result will be:

2023-02-22 21:34:37.472916
The date contains year, month, day, hour, minute, second, and microsecond.

The datetime module has many methods to return information about the date object.

Here are a few examples, you will learn more about them later in this chapter:

In [6]:
# Return the year and name of weekday:

import datetime

x = datetime.datetime.now()
print(x.year)
print(x.strftime("%A"))

2023
Wednesday


Creating Date Objects
--
To create a date, we can use the datetime() class (constructor) of the datetime module.

The datetime() class requires three parameters to create a date: year, month, day.

In [7]:
import datetime

x = datetime.datetime(2020,5,17)
print(x)

2020-05-17 00:00:00


**The datetime() class also takes parameters for time and timezone (hour, minute, second, microsecond, tzone), but they are optional, and has a default value of 0, (None for timezone).**

In [11]:
import datetime

x = datetime.datetime(2020,5,17,23,59,59,245) # with hrs,mins,secs & micro seconds
print(x)

2020-05-17 23:59:59.000245


The strftime() Method
--
The datetime object has a method for formatting date objects into readable strings.

The method is called strftime(), and takes one parameter, format, to specify the format of the returned string:

In [12]:
# Display the name of the month:

import datetime

x = datetime.datetime.now()

print(x.strftime("%B"))

February


In [13]:
# Display the name of the month:

import datetime

x = datetime.datetime(2018,6,24)

print(x.strftime("%B"))

June


In [26]:
import datetime

x = datetime.datetime(2018,6,24)

print(x.strftime("%A"))
print(x.strftime("%B"))
print(x.strftime("%C"))
print(x.strftime("%D"))
# print(x.strftime("%E")) error
print(x.strftime("%F"))
print(x.strftime("%G"))
print(x.strftime("%H"))
print(x.strftime("%I"))
# print(x.strftime("%J")) error
# print(x.strftime("%K"))
#print(x.strftime("%L"))

Sunday
June
20
06/24/18
2018-06-24
2018
00
12
00


In [35]:
import datetime

x = datetime.datetime.now()

print(x.strftime("%A"))
print(x.strftime("%B"))
print(x.strftime("%C"))
print(x.strftime("%D"))
# print(x.strftime("%E"))
print(x.strftime("%F"))
print(x.strftime("%G"))
print(x.strftime("%H"))
print(x.strftime("%I"))
#print(x.strftime("%J"))
#print(x.strftime("%K"))
#print(x.strftime("%L"))

Wednesday
February
20
02/22/23
2023-02-22
2023
21
09


A reference of all the legal format codes:
--

Python Math
--

Python has a set of built-in math functions, including an extensive math module, that allows you to perform mathematical tasks on numbers.



Built-in Math Functions
--
The min() and max() functions can be used to find the lowest or highest value in an iterable:

In [36]:
x = min(5,10,7,2)
y = max(23,34,35,3)

print(x)
print(y)

2
35


**The abs() function returns the absolute (positive) value of the specified number:**

In [37]:
x = abs(-7.25)

print(x)

7.25


**The pow(x, y) function returns the value of x to the power of y (xy).**

In [38]:
# Return the value of 2 to the power of 3 (same as 2 * 2 * 2):

x = pow(2,3)
print(x)

8


The Math Module
--
Python has also a built-in module called math, which extends the list of mathematical functions.

To use it, you must import the math module:

In [39]:
import math

When you have imported the math module, you can start using methods and constants of the module.

The math.sqrt() method for example, returns the square root of a number:

In [40]:
import math

x = dir(math)
print(x)

['__doc__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'comb', 'copysign', 'cos', 'cosh', 'degrees', 'dist', 'e', 'erf', 'erfc', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'isqrt', 'lcm', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'nextafter', 'perm', 'pi', 'pow', 'prod', 'radians', 'remainder', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc', 'ulp']


In [41]:
import math

x = math.sqrt(64)
print(x)

8.0


**The math.ceil() method rounds a number upwards to its nearest integer, and the math.floor() method rounds a number downwards to its nearest integer, and returns the result:**

In [42]:
import math

x = math.ceil(4.2)
y = math.floor(1.9)

print(x)
print(y)

5
1


**The math.pi constant, returns the value of PI (3.14...):**

In [44]:
import math

x = math.pi
print(x)

3.141592653589793


Math Methods
--

Math Constants
--

Python cmath Module
--
Python has a built-in module that you can use for mathematical tasks for complex numbers.

The methods in this module accepts int, float, and complex numbers. It even accepts Python objects that has a __complex__() or __float__() method.

The methods in this module almost always return a complex number. If the return value can be expressed as a real number, the return value has an imaginary part of 0.

The cmath module has a set of methods and constants.

cMath Method
--

cMath Constants
--